### DISCUSSION TITLE
Is it possible to move a search result in the current queue after the current playin song (or specific position)?

### DISCUSSION THREAD

**Discussion by bartfer** [jcorporation/myMPD#997] [Q&A] [answered]
My use case would be that I have a very long queue and sometimes I want to search for a couple of songs and move them to the front of the queue. Like you can do this with only 1 song:
![image](https://user-images.githubusercontent.com/76113978/224481858-e0c46955-1f90-4270-8a91-30f408cff137.png)

Is it possible to somehow execute that command to the search result (multiple songs) in the queue? (not from the DB, I know you can do that from there - which is great!)
Or is it possible to somehow multiselect songs and execute the same commands as for only 1 song?

Sorry if this has been asked before, I couldn't find it.

Thanks!

**Comment by jcorporation** [comment, marked as answer]
This actually not possible. MPD provides no commands to do this. Executing some commands for multiple tracks would be a great enhancement.

**Reply by bartfer** [reply]
Thanks for the quick answer!

Hm, then how does it work when you just search but not in the queue and insert multiple songs after the current playing song? I guess mympd just uses mpc insert command in a for loop, no?
Shouldn't this be just a for loop and executing mpc mv, move <from> <to> - Moves song at position <from> to the position <to> in the queue.
Where TO is 1 or 2, or the current playing number+1?

(or even if there is some problem with mpc move, it is still possible to use mpc insert and then delete the ones that were found in the search)

**Reply by jcorporation** [reply]
myMPD does not use mpc, it uses the mpd protocol (https://mpd.readthedocs.io/en/latest/protocol.html).

Problem with moving multiple songs in the queue is the changing song positions while moving. In my opinion this feature should be implemented on server side. 

Adding a search to the queue works with the mpd command `searchaddpl`.

**Reply by bartfer** [reply]
Sorry if I'm being dumb, but why would that be a problem that your queue is changing song positions? I mean for example:
the search shows the following songs in the queue: 235, 6540, 22142
Then you can - to stay with mpc commands (although as I see mpd protocol commands are the same basically):
mpc mv 235 1
mpc mv 6540 1
mpc mv 22142 1
What would be the problem with this?
I mean if you go in reverse order then yes, you have to add +1 to the sequenceIds.

I see this on the page you linked:
"The position is a 0-based index. It is unstable by design: if you move, delete or insert songs, all following indices will change, and a client can never be sure what song is behind a given index/position."
Which is really, really strange. But I don't see any explanation on what is exactly happening. However I've tried the following commands on my system and the indices seemed correct to me (didn't check all by hand of course :) ). I don't understand what they mean by this. Probably I should ask about this on MPD forum though I guess. 
But if it says adding also messes up indices then this means that searchaddpl function is using a different kind of add? Maybe they implemented that function properly in mpd so that that kind of add doesn't mess up the indices? I'm afraid I'll have to look at their code because this is rather strange. (+also seems to be working anyways)

But as a client another way to solve this could be just deleting the search result songs from the queue and then readding them (using the  `searchaddpl`) after the current playing song. Would this work in mympd?

**Reply by jcorporation** [reply]
The song position is unstable, the song id is stable, as far I know. I do not say that this could not be implemented, I say that it is better implemented on the server side.

It is much work to implement this feature and I am unsure, if it is worth.

The API and the user interface must be extended:

- option to select multiple tracks in queue (have you an idea how this could be implemented for desktop and mobile browsers?)
- extend the API to accept multiple songs
- extend the function to loop through the ids with the command list feature

What functions could benefit from this?

- delete multiple songs from the queue (deleteid)
- play after current song - this involves two methods: moving songs (moveid) around or setting song priority (prioid) for mpd in random mode

This features are more likely implemented, if:
- other myMPD users are also interested in
- you create a feature request for MPD to extend the protocol to allow submitting more than a single id. However I do not know, if it would be accepted.

**Reply by bartfer** [reply]
I think generally multiselecting songs would be a pretty important feature not just for this usecase. You can execute many commands to multiple songs which would make things a lot easier. Just for example adding some songs to a playlist or deletion that you mentioned. Now sometimes I have to fiddle around with "remove all upwards" which usually doesn't cover what I really want to do. One - not too good - way to implement this could be to add an icon to the song row which could mark a song selected. 
(Way better would be if by left clicking the song would not start playing but rather get selected and using shift or ctrl keys you could select multiple songs. I know it's way different on mobile than desktop.)

As for the original question: I highly doubt MPD would implement this because from their point of the view this could be done using the current API.
Thanks for you answer!

**Comment by jcorporation** [comment]
I added a feature request for this. It will be implemented for the v10.4.0 milestone.
